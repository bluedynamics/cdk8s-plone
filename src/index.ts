export { Plone, PloneOptions, PloneBaseOptions, PloneVariant } from './plone';
export { PloneHttpcache, PloneHttpcacheOptions } from './httpcache';
