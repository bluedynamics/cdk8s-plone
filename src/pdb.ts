import { Construct } from 'constructs';
import * as k8s from './imports/k8s';

/**
 * Configuration options for PodDisruptionBudget.
 */
export interface PlonePDBOptions {
  /**
   * Maximum number or percentage of pods that can be unavailable during voluntary disruptions.
   * Can be an absolute number (e.g., 1) or a percentage (e.g., '50%').
   * Cannot be used together with minAvailable.
   * @example 1 or '25%'
   * @default - not set (defaults to minAvailable=1 if neither is specified)
   */
  readonly maxUnavailable?: number | string;

  /**
   * Minimum number or percentage of pods that must remain available during voluntary disruptions.
   * Can be an absolute number (e.g., 1) or a percentage (e.g., '50%').
   * Cannot be used together with maxUnavailable.
   * @example 2 or '50%'
   * @default 1 (if maxUnavailable is not set)
   */
  readonly minAvailable?: number | string;

  /**
   * Additional Kubernetes labels for the PodDisruptionBudget.
   * @default - standard Plone labels only
   */
  readonly labels?: { [name: string]: string };
}

/**
 * PlonePDB creates a Kubernetes PodDisruptionBudget for high availability.
 *
 * This is an internal construct used by PloneDeployment.
 * It ensures a minimum number of pods remain available during voluntary
 * disruptions like node drains or cluster upgrades.
 */
export class PlonePDB extends Construct {

  constructor(scope: Construct, id: string, selectorLabel: { [name: string]: string }, options: PlonePDBOptions) {
    super(scope, id);

    var spec: k8s.PodDisruptionBudgetSpec = {};
    if (typeof options.maxUnavailable === 'number') {
      spec = {
        maxUnavailable: k8s.IntOrString.fromNumber(options.maxUnavailable as number),
      };
    } else if (typeof options.maxUnavailable === 'string') {
      spec = {
        maxUnavailable: k8s.IntOrString.fromString(options.maxUnavailable as string),
      };
    }
    if (typeof options.minAvailable === 'number') {
      spec = {
        ...spec,
        minAvailable: k8s.IntOrString.fromNumber(options.minAvailable as number),
      };
    } else if (typeof options.minAvailable === 'string') {
      spec = {
        ...spec,
        minAvailable: k8s.IntOrString.fromString(options.minAvailable as string),
      };
    }
    if (options.maxUnavailable === undefined && options.minAvailable === undefined) {
      spec = {
        minAvailable: k8s.IntOrString.fromNumber(1),
      };
    }

    spec = {
      ...spec,
      selector: { matchLabels: selectorLabel },
    };
    const pdb_labels = {
      ...options.labels ?? {},
      'app.kubernetes.io/part-of': 'plone',
      'app.kubernetes.io/managed-by': 'cdk8s-plone',
    };

    new k8s.KubePodDisruptionBudget(this, 'PDB', {
      metadata: {
        labels: pdb_labels,
      },
      spec: spec,
    });
  }
}
