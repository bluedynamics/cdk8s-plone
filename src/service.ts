// eslint-disable-next-line import/no-extraneous-dependencies
import { Construct } from 'constructs';
import * as k8s from './imports/k8s';

export interface PloneServiceOptions {

  /**
   * targetPort number.
   */
  readonly targetPort: number;

  /**
   * Selector label.
   */
  readonly selectorLabel: { [name: string]: string };

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };

  /**
   * Port name for the service.
   * @default - 'http'
   */
  readonly portName?: string;
}

/**
 * PloneService creates a Kubernetes Service for accessing Plone pods.
 *
 * This is an internal construct used by the Plone class.
 * It creates a ClusterIP service that routes traffic to the backend
 * or frontend deployment pods.
 */
export class PloneService extends Construct {
  /**
   * The name of the created Kubernetes service.
   * Can be used to reference this service from other resources.
   */
  public name: string;

  constructor(scope: Construct, id: string, options: PloneServiceOptions) {
    super(scope, id);

    const targetPort = k8s.IntOrString.fromNumber(options.targetPort);
    const selectorLabel = options.selectorLabel;
    const service_labels = {
      ...options.labels ?? {},
      'app.kubernetes.io/part-of': 'plone',
      'app.kubernetes.io/managed-by': 'cdk8s-plone',
    };

    const serviceOpts: k8s.KubeServiceProps = {
      metadata: {
        labels: service_labels,
      },
      spec: {
        ports: [{ port: options.targetPort, targetPort: targetPort, name: options.portName ?? 'http' }],
        selector: selectorLabel,
      },
    };
    const service = new k8s.KubeService(this, 'service', serviceOpts);
    this.name = service.name;
  }
}
