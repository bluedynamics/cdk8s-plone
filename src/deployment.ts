// import { log } from 'console';
import { Names } from 'cdk8s';
import * as kplus from 'cdk8s-plus-24';
import { Construct } from 'constructs';
import * as k8s from './imports/k8s';
import { PlonePDB, PlonePDBOptions } from './pdb';

export interface PloneImageOptions {
  /**
   * Specify a custom image for Plone .
   * @default ""
   */
  readonly image?: string;

  /**
   * Specify Secret to pull image .
   * @default ""
   */
  readonly imagePullSecrets?: string[];

  /**
   * Specify Pull Policy .
   * @default ""
   */
  readonly imagePullPolicy?: string;
}

export interface PloneDeploymentOptions {
  /**
   * Specify a custom image for Plone .
   * @default "plone/plone-backend:latest"
   */
  readonly image?: PloneImageOptions;

  /**
   * Specify an environment for Plone .
   * @default - none
  */
  readonly environment?: kplus.Env;

  /**
 * Number of replicas.
 * @default 2
 */
  readonly replicas?: number;

  /**
 * CPU limit
 * @default 1
 */
  readonly limitCpu?: string;

  /**
 * memory limit
 * @default 1
 */
  readonly limitMemory?: string;

  /**
 * CPU request
 * @default 1
 */
  readonly requestCpu?: string;

  /**
 * memory request
 * @default 1
 */
  readonly requestMemory?: string;

  /**
   * Port number.
   */
  readonly port: number;

  /**
   * Extra labels to associate with resources.
   * @default - none
   */
  readonly labels?: { [name: string]: string };

  /**
   * Extra container spec to use for ploencontainer.
   * @default - []
   */
  readonly ploneContainer?: k8s.Container;

  /**
   * Sidecar container spec to associate with resources.
   * @default - []
   */
  readonly sidecars?: k8s.Container[];

  /**
   * Create a PodDisruptionBugdet for the deployment?
   * If given
   * @default - none
   */
  readonly pdb?: PlonePDBOptions;

  /**
   * Liveness Probe for the pod.
   * @default - generated
   */
  readonly livenessProbe?: k8s.Probe;

  /**
   * Readiness Probe for the pod.
   * @default - generated
   */
  readonly readinessProbe?: k8s.Probe;

}

export class PloneDeployment extends Construct {

  constructor(scope: Construct, id: string, options: PloneDeploymentOptions) {
    super(scope, id);
    const image = options.image ?? {};
    const replicas = options.replicas ?? 2;
    const label = { app: Names.toLabelValue(this) };
    const optionLabels = options.labels ?? {};
    const deploymentLabels = {
      'app.kubernetes.io/name': optionLabels['app.kubernetes.io/name'] + '-deployment',
      'app.kubernetes.io/component': optionLabels['app.kubernetes.io/component'] ?? '' + '-deployment',
    };
    const template_labels = {
      ...optionLabels,
      ...label,
      'app.kubernetes.io/part-of': 'plone',
      'app.kubernetes.io/managed-by': 'cdk8s-plone',
    };
    const kpEnv: kplus.Env = options?.environment ?? new kplus.Env([], {});
    var env: k8s.EnvVar[] = [];
    for (const name in kpEnv.variables) {
      env.push({ name: name, value: kpEnv.variables[name].value, valueFrom: kpEnv.variables[name].valueFrom });
    }
    var envFrom: k8s.EnvFromSource[] = [];
    for (const idx in kpEnv.sources) {
      const source = kpEnv.sources[idx];
      envFrom.push(source._toKube());
    }
    var ploneContainerSpec: k8s.Container = {
      name: id + '-container', // here the namespaced name shold be used, but how?
      image: image.image,
      imagePullPolicy: image.imagePullPolicy,
      env: env,
      envFrom: envFrom,
      resources: {
        limits: {
          cpu: k8s.Quantity.fromString(options.limitCpu ?? '1000m'),
          memory: k8s.Quantity.fromString(options.limitMemory ?? '1Gi'),
        },
        requests: {
          cpu: k8s.Quantity.fromString(options.requestCpu ?? '200m'),
          memory: k8s.Quantity.fromString(options.requestMemory ?? '300Mi'),
        },
      },
      livenessProbe: options.livenessProbe ?? {},
      readinessProbe: options.readinessProbe ?? {},
    };
    const deploymentOptions: k8s.KubeDeploymentProps = {
      metadata: {
        labels: deploymentLabels,
      },
      spec: {
        replicas,
        selector: {
          matchLabels: label,
        },
        template: {
          metadata: { labels: template_labels },
          spec: {
            imagePullSecrets: (image.imagePullSecrets ?? []).map((name) => ({ name: name })),
            containers: [
              ploneContainerSpec,
              ...options.sidecars ?? [],
            ],
          },
        },
      },
    };

    new k8s.KubeDeployment(this, 'deployment', deploymentOptions);

    if (options.pdb ?? false) {
      const pdbOptions = options.pdb ?? {};
      new PlonePDB(this, 'pdb', label, pdbOptions);
    }
  }
}
