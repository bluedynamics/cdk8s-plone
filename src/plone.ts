import { Names } from 'cdk8s';
import * as kplus from 'cdk8s-plus-24';
import { Construct } from 'constructs';
import { PloneDeployment } from './deployment';
import { Probe, IntOrString } from './imports/k8s';
import { PloneService } from './service';

export interface PloneBaseOptions {
  readonly image?: string;
  readonly imagePullPolicy?: string;
  readonly replicas?: number;
  readonly maxUnavailable?: number | string;
  readonly minAvailable?: number | string;
  readonly limitCpu?: number;
  readonly limitMemory?: string;
  readonly environment?: kplus.Env;
}
export interface PloneOptions {
  readonly version?: string;
  readonly siteId?: string;
  readonly backend?: PloneBaseOptions;
  readonly frontend?: PloneBaseOptions;
  readonly imagePullSecrets?: string[];
}

export class Plone extends Construct {

  public readonly backendServiceName: string;
  public readonly frontendServiceName: string;

  constructor(scope: Construct, id: string, options: PloneOptions = {}) {
    super(scope, id);

    // ------------------------------------------------------------------------
    // General
    const siteId = options.siteId ?? 'Plone';

    // ------------------------------------------------------------------------
    // Backend
    const backend = options.backend ?? {};
    const backendPort = 8080;
    const backendLabels = {
      'app.kubernetes.io/name': 'plone-backend',
      'app.kubernetes.io/component': 'backend',
      'app.kubernetes.io/version': options.version ?? 'undefined',
    };

    // Probing
    const backendLivenessProbe: Probe = {
      httpGet: {
        path: '/',
        port: IntOrString.fromNumber(backendPort),
      },
      initialDelaySeconds: 30,
      timeoutSeconds: 1,
      periodSeconds: 10,
      successThreshold: 1,
      failureThreshold: 3,
    };
    const backendReadinessProbe: Probe = {
      httpGet: {
        path: '/',
        port: IntOrString.fromNumber(backendPort),
      },
      initialDelaySeconds: 10,
      timeoutSeconds: 15,
      periodSeconds: 10,
      successThreshold: 1,
      failureThreshold: 3,
    };

    // Deployment
    const backendDeployment = new PloneDeployment(this, 'backend', {
      labels: backendLabels,
      image: {
        image: backend.image ?? 'plone/plone-backend:latest',
        imagePullSecrets: options.imagePullSecrets ?? [],
        imagePullPolicy: backend.imagePullPolicy ?? 'IfNotPresent',
      },
      replicas: backend.replicas,
      limitCpu: backend.limitCpu ?? 1,
      limitMemory: backend.limitMemory ?? '512Mi',
      pdb: {
        maxUnavailable: backend.maxUnavailable ?? undefined,
        minAvailable: backend.minAvailable ?? undefined,
      },
      port: backendPort,
      environment: backend.environment,
      livenessProbe: backendLivenessProbe,
      readinessProbe: backendReadinessProbe,
    });

    // Service
    const backendService = new PloneService(backendDeployment, 'service', {
      labels: {
        'app.kubernetes.io/name': 'plone-backend-service',
        'app.kubernetes.io/component': 'service',
      },
      targetPort: backendPort,
      selectorLabel: { app: Names.toLabelValue(backendDeployment) },
    });
    this.backendServiceName = backendService.name;

    // ------------------------------------------------------------------------
    // Frontend
    const frontend = options.frontend ?? {};
    const frontendPort = 3000;
    const frontendLabels = {
      'app.kubernetes.io/name': 'plone-frontend',
      'app.kubernetes.io/component': 'frontend',
      'app.kubernetes.io/version': options.version ?? 'undefined',
    };

    // Probing
    const frontendLivenessProbe: Probe = {
      httpGet: {
        path: '/',
        port: IntOrString.fromNumber(frontendPort),
      },
      initialDelaySeconds: 30,
      timeoutSeconds: 1,
      periodSeconds: 10,
      successThreshold: 1,
      failureThreshold: 3,
    };
    const frontendReadinessProbe: Probe = {
      httpGet: {
        path: '/',
        port: IntOrString.fromNumber(frontendPort),
      },
      initialDelaySeconds: 10,
      timeoutSeconds: 15,
      periodSeconds: 10,
      successThreshold: 1,
      failureThreshold: 3,
    };

    // Environment for RAZZLE
    var frontendEnvironment = frontend.environment ?? new kplus.Env([], {});
    if (frontendEnvironment.variables.RAZZLE_INTERNAL_API_PATH === undefined) {
      // connect with backend service
      frontendEnvironment?.addVariable('RAZZLE_INTERNAL_API_PATH', kplus.EnvValue.fromValue(`http://${backendService.name}:${backendPort}/${siteId}`));
    }

    // Deployment
    const frontendDeployment = new PloneDeployment(this, 'frontend', {
      labels: frontendLabels,
      image: {
        image: frontend.image ?? 'plone/plone-frontend:latest',
        imagePullSecrets: options.imagePullSecrets ?? [],
        imagePullPolicy: frontend.imagePullPolicy ?? 'IfNotPresent',
      },
      replicas: frontend.replicas,
      limitCpu: frontend.limitCpu ?? 1,
      limitMemory: frontend.limitMemory ?? '1Gi',
      pdb: {
        maxUnavailable: frontend.maxUnavailable ?? undefined,
        minAvailable: frontend.minAvailable ?? undefined,
      },
      port: frontendPort,
      environment: frontendEnvironment,
      livenessProbe: frontendLivenessProbe,
      readinessProbe: frontendReadinessProbe,
    });

    // Service
    const frontendService = new PloneService(frontendDeployment, 'service', {
      labels: {
        'app.kubernetes.io/name': 'plone-frontend-service',
        'app.kubernetes.io/component': 'service',
      },
      targetPort: frontendPort,
      selectorLabel: { app: Names.toLabelValue(frontendDeployment) },
    });
    this.frontendServiceName = frontendService.name;
  }
}