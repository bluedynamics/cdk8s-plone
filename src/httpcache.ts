import * as fs from 'fs';
import * as path from 'path';
import { Helm } from 'cdk8s';
import { Construct } from 'constructs';
import { Plone } from './plone';

export interface PloneHttpCacheOptions {
  /**
   * plone chart
   * @default - none
   */
  readonly plone: Plone;

  /** varnishVcl
   * @default - file in config folder
   */
  readonly varnishVcl?: string;

  /** varnishVclFile
   * @default - undefined
   */
  readonly varnishVclFile?: string | undefined;

  /** existingSecret - Read admin credentials from user provided secret
   * @default - undefined
  */
  readonly existingSecret?: string;


  // resources
  readonly limitCpu?: string;
  readonly limitMemory?: string;
  readonly requestCpu?: string;
  readonly requestMemory?: string;
}

export class PloneHttpcache extends Construct {

  public readonly httpcacheServiceName: string;

  constructor(scope: Construct, id: string, options: PloneHttpCacheOptions) {
    super(scope, id);
    let varnishVcl: string;
    if (!options.varnishVcl) {
      let vclPath: string;
      if (!options.varnishVclFile) {
        vclPath = path.join(__dirname, 'config', 'varnish.tpl.vcl');
      } else {
        vclPath = options.varnishVclFile;
      }
      varnishVcl = fs.readFileSync(vclPath, 'utf8');
      throw new Error(varnishVcl);
    } else {
      varnishVcl = options.varnishVcl;
    }
    const httpcache = new Helm(this, 'httpcache', {
      // see https://github.com/mittwald/kube-httpcache/chart
      repo: 'https://helm.mittwald.de',
      chart: 'kube-httpcache',
      values: {
        replicaCount: 2,
        cache: {
          // need to looks at the frontendWatch, do we need it?
          frontendWatch: false,
          backendWatch: false,
          existingSecret: options.existingSecret ?? undefined,
        },
        vclTemplate: varnishVcl,
        extraEnvVars: [
          { name: 'BACKEND_SERVICE_NAME', value: options.plone.backendServiceName },
          { name: 'BACKEND_SERVICE_PORT', value: '8080' },
          { name: 'FRONTEND_SERVICE_NAME', value: options.plone.frontendServiceName },
          { name: 'FRONTEND_SERVICE_PORT', value: '3000' },
        ],
        resources: {
          limits: {
            cpu: options.limitCpu || '500m',
            memory: options.limitMemory || '500Mi',
          },
          requests: {
            cpu: options.requestCpu || '100m',
            memory: options.requestMemory || '100Mi',
          },
        },
        rbac: {
          enabled: false,
        },
        exporter: {
          enabled: true,
          resources: {
            limits: {
              cpu: '100m',
              memory: '100Mi',
            },
            requests: {
              cpu: '10m',
              memory: '50Mi',
            },
          },
        },
        serviceMonitor: {
          enabled: true,
          scrapeSignaller: true,
        },
      },
    });
    const httpcacheService = httpcache.apiObjects.find((construct) => {
      if ((construct.kind === 'Service') && (construct.metadata.name?.endsWith('kube-httpcache'))) {
        return construct.name;
      }
      return undefined;
    });
    if (httpcacheService === undefined) {
      throw new Error('Could not find httpcache service');
    }
    this.httpcacheServiceName = httpcacheService.name;
  }
}