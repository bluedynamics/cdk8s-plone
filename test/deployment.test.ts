import { Chart, Testing } from 'cdk8s';
import * as kplus from 'cdk8s-plus-24';
import { PloneDeployment } from '../src/deployment';

test('defaults', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'plone');

  // WHEN
  new PloneDeployment(chart, 'test', { port: 8080 });

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});

test('with-pdp', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'plone');

  // WHEN
  new PloneDeployment(chart, 'with_pdb', {
    port: 3000,
    pdb: { minAvailable: 1 },
  });

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});

test('with-environment', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'plone');

  // WHEN
  const env = new kplus.Env([], { MY_ENV: { value: 'my-value' } });
  new PloneDeployment(chart, 'with_environment', {
    port: 3000,
    environment: env,
  });

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});

test('with-environment-valueFrom', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'plone');

  // WHEN
  const env = new kplus.Env([], { MY_ENV: { valueFrom: { secretKeyRef: { name: 'my-secret', key: 'MY_ENV_SECRET' } } } });
  new PloneDeployment(chart, 'with_environment_valueFrom', {
    port: 3000,
    environment: env,
  });

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});

test('with-environment-from-secret', () => {
  // GIVEN
  const app = Testing.app();
  const chart = new Chart(app, 'plone');

  // WHEN
  const envfrom = kplus.Env.fromSecret(new kplus.Secret(chart, 'foo'));
  const env = new kplus.Env([envfrom], {});
  new PloneDeployment(chart, 'with_environment_secret', {
    port: 3000,
    environment: env,
  });

  // THEN
  expect(Testing.synth(chart)).toMatchSnapshot();
});
