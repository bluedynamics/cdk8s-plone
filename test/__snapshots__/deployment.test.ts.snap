// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults 1`] = `
[
  {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": {
      "labels": {
        "app.kubernetes.io/component": "-deployment",
        "app.kubernetes.io/name": "undefined-deployment",
      },
      "name": "plone-test-deployment-c86dcee1",
    },
    "spec": {
      "replicas": 2,
      "selector": {
        "matchLabels": {
          "app": "plone-test-c8667035",
        },
      },
      "template": {
        "metadata": {
          "labels": {
            "app": "plone-test-c8667035",
            "app.kubernetes.io/managed-by": "cdk8s-plone",
            "app.kubernetes.io/part-of": "plone",
          },
        },
        "spec": {
          "containers": [
            {
              "env": [],
              "envFrom": [],
              "livenessProbe": {},
              "name": "test-container",
              "readinessProbe": {},
              "resources": {
                "limits": {
                  "cpu": 1,
                  "memory": "1Gi",
                },
              },
            },
          ],
          "imagePullSecrets": [],
        },
      },
    },
  },
]
`;

exports[`with-environment 1`] = `
[
  {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": {
      "labels": {
        "app.kubernetes.io/component": "-deployment",
        "app.kubernetes.io/name": "undefined-deployment",
      },
      "name": "plone-with_environment-deployment-c80b2568",
    },
    "spec": {
      "replicas": 2,
      "selector": {
        "matchLabels": {
          "app": "plone-with_environment-c8045a0c",
        },
      },
      "template": {
        "metadata": {
          "labels": {
            "app": "plone-with_environment-c8045a0c",
            "app.kubernetes.io/managed-by": "cdk8s-plone",
            "app.kubernetes.io/part-of": "plone",
          },
        },
        "spec": {
          "containers": [
            {
              "env": [
                {
                  "name": "MY_ENV",
                  "value": "my-value",
                },
              ],
              "envFrom": [],
              "livenessProbe": {},
              "name": "with_environment-container",
              "readinessProbe": {},
              "resources": {
                "limits": {
                  "cpu": 1,
                  "memory": "1Gi",
                },
              },
            },
          ],
          "imagePullSecrets": [],
        },
      },
    },
  },
]
`;

exports[`with-environment-from-secret 1`] = `
[
  {
    "apiVersion": "v1",
    "immutable": false,
    "kind": "Secret",
    "metadata": {
      "name": "plone-foo-c8953a75",
    },
    "stringData": {},
  },
  {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": {
      "labels": {
        "app.kubernetes.io/component": "-deployment",
        "app.kubernetes.io/name": "undefined-deployment",
      },
      "name": "plone-with_environment_secret-deployment-c8c68046",
    },
    "spec": {
      "replicas": 2,
      "selector": {
        "matchLabels": {
          "app": "plone-with_environment_secret-c8cb976e",
        },
      },
      "template": {
        "metadata": {
          "labels": {
            "app": "plone-with_environment_secret-c8cb976e",
            "app.kubernetes.io/managed-by": "cdk8s-plone",
            "app.kubernetes.io/part-of": "plone",
          },
        },
        "spec": {
          "containers": [
            {
              "env": [],
              "envFrom": [
                {
                  "secretRef": {
                    "name": "plone-foo-c8953a75",
                  },
                },
              ],
              "livenessProbe": {},
              "name": "with_environment_secret-container",
              "readinessProbe": {},
              "resources": {
                "limits": {
                  "cpu": 1,
                  "memory": "1Gi",
                },
              },
            },
          ],
          "imagePullSecrets": [],
        },
      },
    },
  },
]
`;

exports[`with-environment-valueFrom 1`] = `
[
  {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": {
      "labels": {
        "app.kubernetes.io/component": "-deployment",
        "app.kubernetes.io/name": "undefined-deployment",
      },
      "name": "plone-with_environment_valuefrom-deployment-c8068bfb",
    },
    "spec": {
      "replicas": 2,
      "selector": {
        "matchLabels": {
          "app": "plone-with_environment_valueFrom-c89a569b",
        },
      },
      "template": {
        "metadata": {
          "labels": {
            "app": "plone-with_environment_valueFrom-c89a569b",
            "app.kubernetes.io/managed-by": "cdk8s-plone",
            "app.kubernetes.io/part-of": "plone",
          },
        },
        "spec": {
          "containers": [
            {
              "env": [
                {
                  "name": "MY_ENV",
                  "valueFrom": {
                    "secretKeyRef": {
                      "key": "MY_ENV_SECRET",
                      "name": "my-secret",
                    },
                  },
                },
              ],
              "envFrom": [],
              "livenessProbe": {},
              "name": "with_environment_valueFrom-container",
              "readinessProbe": {},
              "resources": {
                "limits": {
                  "cpu": 1,
                  "memory": "1Gi",
                },
              },
            },
          ],
          "imagePullSecrets": [],
        },
      },
    },
  },
]
`;

exports[`with-pdp 1`] = `
[
  {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": {
      "labels": {
        "app.kubernetes.io/component": "-deployment",
        "app.kubernetes.io/name": "undefined-deployment",
      },
      "name": "plone-with_pdb-deployment-c87b10de",
    },
    "spec": {
      "replicas": 2,
      "selector": {
        "matchLabels": {
          "app": "plone-with_pdb-c80f43fe",
        },
      },
      "template": {
        "metadata": {
          "labels": {
            "app": "plone-with_pdb-c80f43fe",
            "app.kubernetes.io/managed-by": "cdk8s-plone",
            "app.kubernetes.io/part-of": "plone",
          },
        },
        "spec": {
          "containers": [
            {
              "env": [],
              "envFrom": [],
              "livenessProbe": {},
              "name": "with_pdb-container",
              "readinessProbe": {},
              "resources": {
                "limits": {
                  "cpu": 1,
                  "memory": "1Gi",
                },
              },
            },
          ],
          "imagePullSecrets": [],
        },
      },
    },
  },
  {
    "apiVersion": "policy/v1",
    "kind": "PodDisruptionBudget",
    "metadata": {
      "labels": {
        "app.kubernetes.io/managed-by": "cdk8s-plone",
        "app.kubernetes.io/part-of": "plone",
      },
      "name": "plone-with_pdb-pdb-c8380749",
    },
    "spec": {
      "minAvailable": 1,
      "selector": {
        "matchLabels": {
          "app": "plone-with_pdb-c80f43fe",
        },
      },
    },
  },
]
`;
